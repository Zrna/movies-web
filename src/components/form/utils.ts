import { FieldMetaState } from 'react-final-form';

import { TextInputProps } from '~/ui';

export interface FormFieldProps extends Omit<TextInputProps, 'value' | 'onChange'> {
  name: string;
  type?: string;
  label?: string;
  validate?: any;
  value?: null;
  onChange?(): null;
}

export function getFieldError(meta: FieldMetaState<any>) {
  if (meta.touched && meta.error) {
    return meta.error;
  }

  return meta.submitError;
}
