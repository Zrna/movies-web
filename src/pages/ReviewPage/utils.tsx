import { theme } from '~/ui';

export const reactionsMap: { [key: number]: { icon: theme.Icon; text: string } } = {
  0: { icon: 'reaction0', text: 'No comment' },
  1: { icon: 'reaction1', text: "I wish I hadn't watched" },
  2: { icon: 'reaction2', text: 'Boring' },
  3: { icon: 'reaction3', text: 'Not great, not terrible' },
  4: { icon: 'reaction4', text: "It's great" },
  5: { icon: 'reaction5', text: 'Loving it!' },
};
