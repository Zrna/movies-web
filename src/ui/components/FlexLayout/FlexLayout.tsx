import { Property } from 'csstype';
import { forwardRef } from 'react';

import { Box, BoxProps, useResponsiveSelector } from '~/ui';

export interface FlexLayoutProps extends BoxProps {
  alignItems?: Property.AlignItems | Property.AlignItems[];
  flexDirection?: Property.FlexDirection | Property.FlexDirection[];
  flexGrow?: Property.FlexGrow | Property.FlexGrow[];
  flexShrink?: Property.FlexShrink | Property.FlexShrink[];
  flexWrap?: Property.FlexWrap | Property.FlexWrap[];
  justifyContent?: Property.JustifyContent | Property.JustifyContent[];
  space?: number | number[];
}

type PossiblyArray<T> = T | T[];

function mapArrayOrOne<T, V>(items: PossiblyArray<T>, f: (value: T) => V) {
  return Array.isArray(items) ? items.map(f) : f(items);
}

export const FlexLayout = forwardRef<any, FlexLayoutProps>(
  (
    { alignItems, flexDirection = 'row', flexGrow, flexShrink, flexWrap, justifyContent, space = 0, sx = {}, ...rest },
    ref,
  ) => {
    const r = useResponsiveSelector();

    const spaceValue = Array.isArray(space) ? r(...space) : space;

    const allButLast = {
      mb: mapArrayOrOne(flexDirection, (d) => (d === 'column' ? spaceValue : 0)),
      mr: mapArrayOrOne(flexDirection, (d) => (d === 'row' ? spaceValue : 0)),
    };

    return (
      <Box
        ref={ref}
        sx={{
          ...sx,
          alignItems,
          display: 'flex',
          flexDirection,
          flexGrow,
          flexShrink,
          flexWrap,
          justifyContent,
          '> :not(:last-child)': allButLast,
        }}
        {...rest}
      />
    );
  },
);
