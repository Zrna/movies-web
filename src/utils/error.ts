import { AxiosError } from 'axios';

import { showToast } from '~/ui';

export function getErrorMessage(error: AxiosError) {
  if (error.response) {
    return error.response.data.error;
  }

  return 'Something went wrong';
}

export function showErrorToast(error: AxiosError) {
  showToast({
    variant: 'error',
    text: getErrorMessage(error),
  });
}
