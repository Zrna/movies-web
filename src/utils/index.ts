export * from './async';
export * from './error';
export * as validator from './validation';
