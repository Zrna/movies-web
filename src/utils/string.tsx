import { Fragment } from 'react';

export function splitStringToNewLine(str: string) {
  return str.split('\n').map((value, i) => (
    <Fragment key={value + i}>
      {value} <br />
    </Fragment>
  ));
}

export function pluralize(noun: string, length: number) {
  return length === 1 ? noun : `${noun}s`;
}
