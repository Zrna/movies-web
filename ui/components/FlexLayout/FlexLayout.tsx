import { Property } from 'csstype';
import { forwardRef } from 'react';

import { Box, BoxProps } from '~/ui';

export interface FlexLayoutProps extends BoxProps {
  alignItems?: Property.AlignItems | Property.AlignItems[];
  flexDirection?: Property.FlexDirection | Property.FlexDirection[];
  flexGrow?: Property.FlexGrow | Property.FlexGrow[];
  flexShrink?: Property.FlexShrink | Property.FlexShrink[];
  flexWrap?: Property.FlexWrap | Property.FlexWrap[];
  justifyContent?: Property.JustifyContent | Property.JustifyContent[];
  space?: number | number[];
}

export const FlexLayout = forwardRef<any, FlexLayoutProps>(
  (
    { alignItems, flexDirection = 'row', flexGrow, flexShrink, flexWrap, justifyContent, space = 0, sx = {}, ...rest },
    ref,
  ) => {
    return (
      <Box
        ref={ref}
        sx={{
          ...sx,
          alignItems,
          display: 'flex',
          flexDirection,
          flexGrow,
          flexShrink,
          flexWrap,
          justifyContent,
          gap: space,
          '> :not(:last-child)': '', //TODO: allButLast,
        }}
        {...rest}
      />
    );
  },
);
