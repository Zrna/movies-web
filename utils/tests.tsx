import { cleanup, fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { ThemeProvider } from 'theme-ui';

import { theme } from '~/ui';

export { cleanup, fireEvent, render, userEvent };

export function renderWithTheme(Component: JSX.Element) {
  return render(<ThemeProvider theme={theme}>{Component}</ThemeProvider>);
}

export function getThemeColorString(color: theme.Color) {
  return `var(--theme-ui-colors-${color})`;
}

export function expectTextVariant(element: HTMLElement, variant: theme.TextVariant) {
  const { fontFamily, fontSize, fontStyle, fontWeight, lineHeight } = theme.text[variant];

  expect(element).toHaveStyleRule('font-family', fontFamily);
  expect(element).toHaveStyleRule('font-style', fontStyle);
  expect(element).toHaveStyleRule('font-size', fontSize);
  expect(element).toHaveStyleRule('font-weight', fontWeight.toString());
  expect(element).toHaveStyleRule('line-height', lineHeight);

  if (variant === 'label-uppercase') {
    expect(element).toHaveStyleRule('text-transform', theme.text[variant].textTransform);
  }
}

export function mockWindowMatchMedia() {
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });
}
