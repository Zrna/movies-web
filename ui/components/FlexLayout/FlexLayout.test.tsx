import { screen, within } from '@testing-library/react';

import { mockWindowMatchMedia, renderWithTheme } from '~/utils/tests';

import { FlexLayout } from './FlexLayout';

describe('FlexLayout', () => {
  beforeAll(mockWindowMatchMedia);

  it('should render', () => {
    renderWithTheme(
      <FlexLayout
        alignItems="center"
        data-testid="flex"
        flexDirection="column"
        flexGrow={1}
        flexWrap="wrap"
        justifyContent="space-between"
        space={4}
      >
        <div>Item 1</div>
        <div>Item 2</div>
        <div>Item 3</div>
      </FlexLayout>,
    );

    expect(screen.getByTestId('flex')).toMatchSnapshot();
  });

  it('should render flex component', () => {
    renderWithTheme(<FlexLayout data-testid="flex" />);

    expect(screen.getByTestId('flex')).toHaveStyle({ display: 'flex' });
  });

  describe('flex container', () => {
    it('should apply proper direction', () => {
      renderWithTheme(<FlexLayout data-testid="flex1" flexDirection="row" />);
      renderWithTheme(<FlexLayout data-testid="flex2" flexDirection="column" />);
      renderWithTheme(<FlexLayout data-testid="flex3" flexDirection="row-reverse" />);
      renderWithTheme(<FlexLayout data-testid="flex4" flexDirection="column-reverse" />);

      expect(screen.getByTestId('flex1')).toHaveStyle({ flexDirection: 'row' });
      expect(screen.getByTestId('flex2')).toHaveStyle({ flexDirection: 'column' });
      expect(screen.getByTestId('flex3')).toHaveStyle({ flexDirection: 'row-reverse' });
      expect(screen.getByTestId('flex4')).toHaveStyle({ flexDirection: 'column-reverse' });
    });

    it('should align items properly', () => {
      renderWithTheme(<FlexLayout alignItems="center" data-testid="flex1" />);
      renderWithTheme(<FlexLayout alignItems="flex-start" data-testid="flex2" />);
      renderWithTheme(<FlexLayout alignItems="flex-end" data-testid="flex3" />);
      renderWithTheme(<FlexLayout alignItems="stretch" data-testid="flex4" />);
      renderWithTheme(<FlexLayout alignItems="baseline" data-testid="flex5" />);

      expect(screen.getByTestId('flex1')).toHaveStyle({ alignItems: 'center' });
      expect(screen.getByTestId('flex2')).toHaveStyle({ alignItems: 'flex-start' });
      expect(screen.getByTestId('flex3')).toHaveStyle({ alignItems: 'flex-end' });
      expect(screen.getByTestId('flex4')).toHaveStyle({ alignItems: 'stretch' });
      expect(screen.getByTestId('flex5')).toHaveStyle({ alignItems: 'baseline' });
    });

    it('should justify items properly', () => {
      renderWithTheme(<FlexLayout data-testid="flex1" justifyContent="center" />);
      renderWithTheme(<FlexLayout data-testid="flex2" justifyContent="flex-start" />);
      renderWithTheme(<FlexLayout data-testid="flex3" justifyContent="flex-end" />);
      renderWithTheme(<FlexLayout data-testid="flex4" justifyContent="space-between" />);
      renderWithTheme(<FlexLayout data-testid="flex5" justifyContent="space-around" />);
      renderWithTheme(<FlexLayout data-testid="flex6" justifyContent="space-evenly" />);

      expect(screen.getByTestId('flex1')).toHaveStyle({ justifyContent: 'center' });
      expect(screen.getByTestId('flex2')).toHaveStyle({ justifyContent: 'flex-start' });
      expect(screen.getByTestId('flex3')).toHaveStyle({ justifyContent: 'flex-end' });
      expect(screen.getByTestId('flex4')).toHaveStyle({ justifyContent: 'space-between' });
      expect(screen.getByTestId('flex5')).toHaveStyle({ justifyContent: 'space-around' });
      expect(screen.getByTestId('flex6')).toHaveStyle({ justifyContent: 'space-evenly' });
    });

    it('should wrap items properly', () => {
      renderWithTheme(<FlexLayout data-testid="flex1" flexWrap="wrap" />);
      renderWithTheme(<FlexLayout data-testid="flex2" flexWrap="wrap-reverse" />);
      renderWithTheme(<FlexLayout data-testid="flex3" flexWrap="nowrap" />);

      expect(screen.getByTestId('flex1')).toHaveStyle({ flexWrap: 'wrap' });
      expect(screen.getByTestId('flex2')).toHaveStyle({ flexWrap: 'wrap-reverse' });
      expect(screen.getByTestId('flex3')).toHaveStyle({ flexWrap: 'nowrap' });
    });

    it('should apply proper grow and shrink', () => {
      renderWithTheme(<FlexLayout data-testid="flex1" flexGrow={2} flexShrink={1} />);
      renderWithTheme(<FlexLayout data-testid="flex2" flexGrow={1} flexShrink={0} />);

      expect(screen.getByTestId('flex1')).toHaveStyle({ flexGrow: '2', flexShrink: 1 });
      expect(screen.getByTestId('flex2')).toHaveStyle({ flexGrow: '1', flexShrink: 0 });
    });
  });

  describe('child items', () => {
    it('should render child items', () => {
      renderWithTheme(
        <FlexLayout data-testid="flex">
          <div>Item 1</div>
          <div>Item 2</div>
          <div>Item 3</div>
        </FlexLayout>,
      );

      const flexElement = screen.getByTestId('flex');

      expect(within(flexElement).getByText('Item 1')).toBeInTheDocument();
      expect(within(flexElement).getByText('Item 2')).toBeInTheDocument();
      expect(within(flexElement).getByText('Item 3')).toBeInTheDocument();
    });
  });

  describe('spacing', () => {
    it('should apply the gap property', () => {
      renderWithTheme(
        <FlexLayout data-testid="flex" space={4}>
          <div>Item 1</div>
          <div>Item 2</div>
          <div>Item 3</div>
        </FlexLayout>,
      );

      const flexElement = screen.getByTestId('flex');

      expect(flexElement).toHaveStyle({ gap: '16px' });
    });
  });
});
