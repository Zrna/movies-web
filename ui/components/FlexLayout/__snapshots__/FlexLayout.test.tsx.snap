// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlexLayout should render 1`] = `
<div
  class="css-13e4cdg-Box"
  data-testid="flex"
>
  <div>
    Item 1
  </div>
  <div>
    Item 2
  </div>
  <div>
    Item 3
  </div>
</div>
`;
